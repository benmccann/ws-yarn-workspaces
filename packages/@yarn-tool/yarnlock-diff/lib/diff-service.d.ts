/// <reference types="node" />
import { Diff } from "deep-diff";
import { Option } from "fp-ts/lib/Option";
import { ITSValueOrArray } from 'ts-type/lib/type/base';
import { IComputedPackage } from './diff-service/types';
export declare function buildDiff(oldYarnLockContent: ITSValueOrArray<Buffer | string>, newYarnLockContent: ITSValueOrArray<Buffer | string>): Option<Diff<IComputedPackage, IComputedPackage>[]>;
