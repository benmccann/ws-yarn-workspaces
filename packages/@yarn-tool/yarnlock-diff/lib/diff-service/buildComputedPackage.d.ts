/// <reference types="node" />
import { IComputedPackage } from './types';
import { ITSValueOrArray } from 'ts-type/lib/type/base';
export declare function buildComputedPackage(yarnLockContentList: ITSValueOrArray<Buffer | string>, alreadyComputedPackage?: IComputedPackage): IComputedPackage;
