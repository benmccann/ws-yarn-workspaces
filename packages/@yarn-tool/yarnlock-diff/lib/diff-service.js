"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildDiff = void 0;
const deep_diff_1 = require("@bluelovers/deep-diff");
const buildComputedPackage_1 = require("./diff-service/buildComputedPackage");
function buildDiff(oldYarnLockContent, newYarnLockContent) {
    var _a;
    const oldPacakges = buildComputedPackage_1.buildComputedPackage(oldYarnLockContent);
    const newPackages = buildComputedPackage_1.buildComputedPackage(newYarnLockContent);
    return (_a = deep_diff_1.diff(oldPacakges, newPackages)) !== null && _a !== void 0 ? _a : [];
}
exports.buildDiff = buildDiff;
//# sourceMappingURL=diff-service.js.map