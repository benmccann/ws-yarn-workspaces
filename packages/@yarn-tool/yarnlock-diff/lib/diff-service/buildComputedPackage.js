"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildComputedPackage = buildComputedPackage;
const yarnlock_parse_1 = require("@yarn-tool/yarnlock-parse");
const computeHashmapOfPackageAndVersionList_1 = require("./computeHashmapOfPackageAndVersionList");
function buildComputedPackage(yarnLockContentList, alreadyComputedPackage = {}) {
    if (!Array.isArray(yarnLockContentList)) {
        yarnLockContentList = [yarnLockContentList];
    }
    return yarnLockContentList
        .map(v => (0, yarnlock_parse_1.yarnLockParse)(v))
        .reduce(computeHashmapOfPackageAndVersionList_1.computeHashmapOfPackageAndVersionList, alreadyComputedPackage);
}
//# sourceMappingURL=buildComputedPackage.js.map