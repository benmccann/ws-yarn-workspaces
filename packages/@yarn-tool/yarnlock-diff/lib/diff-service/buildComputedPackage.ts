import { IComputedPackage } from './types';
import { yarnLockParse } from '@yarn-tool/yarnlock-parse/index';
import { computeHashmapOfPackageAndVersionList } from './computeHashmapOfPackageAndVersionList';
import { ITSValueOrArray } from 'ts-type/lib/type/base';

export function buildComputedPackage(yarnLockContentList: ITSValueOrArray<Buffer | string>, alreadyComputedPackage: IComputedPackage = {})
{
	if (!Array.isArray(yarnLockContentList))
	{
		yarnLockContentList = [yarnLockContentList];
	}

	return yarnLockContentList
		.map(v => yarnLockParse(v))
	.reduce(computeHashmapOfPackageAndVersionList, alreadyComputedPackage);
}
