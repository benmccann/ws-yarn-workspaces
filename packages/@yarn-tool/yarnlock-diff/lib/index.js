"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.YarnLockDiff = void 0;
const command_1 = require("@oclif/command");
const formatter_1 = require("./formatter");
const diff_service_1 = require("./diff-service");
const path_1 = require("path");
const fs_1 = require("fs");
class YarnLockDiff extends command_1.Command {
    constructor() {
        super(...arguments);
        this.checkIfExist = (filePathString) => {
            const filePath = path_1.resolve(process.cwd(), filePathString);
            return fs_1.existsSync(filePath);
        };
        this.reportNonExistantFiles = (filePathString) => {
            if (!this.checkIfExist(filePathString)) {
                this.error(`File does not exist "${filePathString}"`);
            }
            return filePathString;
        };
        this.checkIfLockFile = (filePathString) => {
            const fileExtension = path_1.extname(filePathString);
            return fileExtension === ".lock";
        };
        this.reportNonLockFile = (filePathString) => {
            if (!this.checkIfLockFile(filePathString)) {
                this.error(`File is not a lock file "${filePathString}"`);
            }
            return filePathString;
        };
    }
    async run() {
        const { flags } = this.parse(YarnLockDiff);
        const oldYarnLockPaths = flags.old;
        const newYarnLockPaths = flags.new;
        [...oldYarnLockPaths, ...newYarnLockPaths]
            .map(this.reportNonExistantFiles)
            .map(this.reportNonLockFile);
        const oldContents = oldYarnLockPaths.map(path => fs_1.readFileSync(path, "utf8"));
        const newContents = newYarnLockPaths.map(path => fs_1.readFileSync(path, "utf8"));
        diff_service_1.buildDiff(oldContents, newContents)
            .map(formatter_1.buildDiffTable)
            .map(this.log);
    }
}
exports.YarnLockDiff = YarnLockDiff;
YarnLockDiff.description = "Given one or more old yarn.lock files and one or more new yarn.lock files, compute the diff";
YarnLockDiff.flags = {
    version: command_1.flags.version({ char: "v" }),
    help: command_1.flags.help({ char: "h" }),
    old: command_1.flags.string({
        char: "o",
        description: "old yarn.lock file(s)",
        multiple: true,
        required: true,
    }),
    new: command_1.flags.string({
        char: "n",
        description: "new yarn.lock file(s)",
        multiple: true,
        required: true,
    }),
};
exports.default = YarnLockDiff;
//# sourceMappingURL=index.js.map