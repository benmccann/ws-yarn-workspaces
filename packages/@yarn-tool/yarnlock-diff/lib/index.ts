import { Command, flags } from "@oclif/command";

import { buildDiffTable } from './formatter';
import { buildDiff } from './diff-service';
import { extname, resolve } from 'path';
import { existsSync, readFileSync } from 'fs';

export class YarnLockDiff extends Command
{
	static description =
		"Given one or more old yarn.lock files and one or more new yarn.lock files, compute the diff";

	static flags = {
		version: flags.version({ char: "v" }),
		help: flags.help({ char: "h" }),
		old: flags.string({
			char: "o",
			description: "old yarn.lock file(s)",
			multiple: true,
			required: true,
		}),
		new: flags.string({
			char: "n",
			description: "new yarn.lock file(s)",
			multiple: true,
			required: true,
		}),
	};

	checkIfExist = (filePathString: string) =>
	{
		const filePath = resolve(process.cwd(), filePathString);
		return existsSync(filePath);
	};

	reportNonExistantFiles = (filePathString: string) =>
	{
		if (!this.checkIfExist(filePathString))
		{
			this.error(`File does not exist "${filePathString}"`);
		}
		return filePathString;
	};

	checkIfLockFile = (filePathString: string) =>
	{
		const fileExtension = extname(filePathString);
		return fileExtension === ".lock";
	};

	reportNonLockFile = (filePathString: string) =>
	{
		if (!this.checkIfLockFile(filePathString))
		{
			this.error(`File is not a lock file "${filePathString}"`);
		}
		return filePathString;
	};

	async run()
	{
		const { flags } = this.parse(YarnLockDiff);

		const oldYarnLockPaths = flags.old;
		const newYarnLockPaths = flags.new;

		[...oldYarnLockPaths, ...newYarnLockPaths]
			.map(this.reportNonExistantFiles)
			.map(this.reportNonLockFile);

		const oldContents = oldYarnLockPaths.map(path =>
			readFileSync(path, "utf8"),
		);
		const newContents = newYarnLockPaths.map(path =>
			readFileSync(path, "utf8"),
		);

		buildDiff(oldContents, newContents)
			.map(buildDiffTable)
			.map(this.log);
	}
}

export default YarnLockDiff;
