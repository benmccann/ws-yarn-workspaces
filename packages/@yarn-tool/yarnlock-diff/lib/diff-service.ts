import { diff, Diff } from "deep-diff";
import { fromNullable, Option } from "fp-ts/lib/Option";
import { yarnLockParse } from '@yarn-tool/yarnlock-parse/index';
import { computeHashmapOfPackageAndVersionList } from './diff-service/computeHashmapOfPackageAndVersionList';
import { buildComputedPackage } from './diff-service/buildComputedPackage';
import { ITSValueOrArray } from 'ts-type/lib/type/base';
import { IComputedPackage } from './diff-service/types';

export function buildDiff(
	oldYarnLockContent: ITSValueOrArray<Buffer | string>,
	newYarnLockContent: ITSValueOrArray<Buffer | string>,
): Option<Diff<IComputedPackage, IComputedPackage>[]>
{
	const oldPacakges = buildComputedPackage(oldYarnLockContent);

	const newPackages = buildComputedPackage(newYarnLockContent);

	return fromNullable(diff(oldPacakges, newPackages));
}
