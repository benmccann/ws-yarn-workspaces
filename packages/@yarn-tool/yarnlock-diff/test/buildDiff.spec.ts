import FastGlob from '@bluelovers/fast-glob';
import { join } from "path";
import { readFileSync } from "fs";
import { buildDiff } from '../lib/diff-service';
import { __TEST_YARNLOCK } from '../../../../__root_ws';

describe(`buildDiff`, () =>
{
	const __res = join(__TEST_YARNLOCK, 'fixtures');

	const files = FastGlob.sync([
			'*.lock',
		], {
			cwd: join(__res, 'a'),
		})
	;

	files.forEach(file =>
	{

		test(file, () =>
		{
			const yarnlock_old = readFileSync(join(__res, 'a', file));
			const yarnlock_new = readFileSync(join(__res, 'b', file));

			let actual = buildDiff(yarnlock_old, yarnlock_new);

			expect(actual).toMatchSnapshot();

		});

	})
	;

})
