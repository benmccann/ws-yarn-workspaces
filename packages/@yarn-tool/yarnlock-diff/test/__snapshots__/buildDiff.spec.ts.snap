// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildDiff yarn.lock 1`] = `
Object {
  "_tag": "Some",
  "value": Array [
    DiffEdit {
      "kind": "E",
      "lhs": "7.0.2",
      "path": Array [
        "@babel/code-frame",
        0,
      ],
      "rhs": "7.0.0",
    },
    DiffEdit {
      "kind": "E",
      "lhs": "7.3.5",
      "path": Array [
        "@babel/generator",
        1,
      ],
      "rhs": "7.3.4",
    },
    DiffEdit {
      "kind": "E",
      "lhs": "6.3.7",
      "path": Array [
        "@babel/generator",
        0,
      ],
      "rhs": "6.3.4",
    },
    DiffDeleted {
      "kind": "D",
      "lhs": Array [
        "7.3.4",
      ],
      "path": Array [
        "@babel/types",
      ],
    },
    DiffNew {
      "kind": "N",
      "path": Array [
        "@babel/parser",
      ],
      "rhs": Array [
        "7.3.4",
      ],
    },
  ],
}
`;
